### PR TITLE
Tessa/select with label

### PR DESCRIPTION
**Pull request by tesk9**
Fixes https://github.com/NoRedInk/noredink-ui/issues/680. 

Relevant to work in https://github.com/NoRedInk/NoRedInk/pull/36205 (I was really struggling to get the elements to line up, and I think the problem stemmed from inconsistent labels in noredink-ui!)

Adds more fully-featured Select.V8 that includes a label.

Screen Shot 2021-11-04 at 12 51 52 PM

@NoRedInk/design 

---

For reviewer -- I suggest ignoring the first commit! it's a straight copy, no changes. https://github.com/NoRedInk/noredink-ui/pull/773/files/efa89877ec90ee2f780cef21b7f7c51d270d0118..2e7a95a8224a222112ecc286818d77424ecfffa8

### CODE REVIEW

**Comment by bendansby** [comment]
El Dorado and Atlantis all rolled into one!

File: src/Nri/Ui/Select/V8.elm [hunk 1]
```diff
@@ -12,10 +12,11 @@ module Nri.Ui.Select.V8 exposing
 {-| Build a select input with a label, optional guidance, and error messaging.
 
 
-# Changes from V5
+# Changes from V7
 
     - view adds a label
     - adds standard custom, nriDescription, etc. attributes
+    - switches to a list-based attribuet API from a record-based API
```
Comment (LearningNerd) [inline src/Nri/Ui/Select/V8.elm:19]: tiny tiny nitpick, typo "attribuet"

**Review by LearningNerd — APPROVED** [review]
Whew, what an improvement! I love all the extra helper functions (even one for the container CSS, great idea), the separate modules for `InputErrorInternal` and `InputLabelInternal`, the built-in `Select.errorMessage`!  🤯😍  

Ran it locally and loving the updated example in the style guide app too. All looks good!

I just posted an inline question about the `value` helper and related updates in that commit, that one confused me the most lol.

File: src/Nri/Ui/Select/V8.elm [hunk 2]
```diff
@@ -58,51 +61,59 @@ we'll automatically generate one from the label you pass in, but this can
 cause problems if you have more than one text input with the same label on
 the page. Use this to be more specific and avoid issues with duplicate IDs!
 -}
-id : String -> Attribute
+id : String -> Attribute value
```
[the remark this answers is not in the document]
Comment (tesk9) [reply src/Nri/Ui/Select/V8.elm]: This is another instance of a type hole!

Consider:
```
module SomeModule exposing (OpaqueType, opaqueTypeConstructor)

type OpaqueType 
    = OpaqueType {  value : String }
    
opaqueTypeConstructor : String -> OpaqueType
opaqueTypeConstructor string =
    OpaqueType { value = string } 
```

As long as the value wrapped by `OpaqueType` is always `{ value : String}`, we know _exactly_ what the type must be in all cases -- there's nothing variable in the type.

However, as soon as we say that the value might _not_ be a `String`, that `value` could be anything at all, we've introduced a type hole.

```
module SomeModule exposing (OpaqueType, opaqueTypeConstructor)

type OpaqueType value
    = OpaqueType {  value : value }
    
opaqueTypeConstructor : value -> OpaqueType value
opaqueTypeConstructor someValue =
    OpaqueType { value = someValue } 
```

Elm requires us to say explicitly when a type values, so we add a type variable to the annotations.

Does that make sense?

[the remark this answers is not in the document]
Comment (LearningNerd) [reply src/Nri/Ui/Select/V8.elm]: Pretty late returning to the party here, but YES it makes a lot more sense now that I'm looking at all of this again. :) thank you for the examples!!

**Pull request merged**